### PR TITLE
CCI: New LDV Data Factory Recipe & Flow

### PR DESCRIPTION
**Pull request by force2b**
- Create a new CCI Flow with associated Snowfakery recipe to insert up to millions of rows into a managed org. 
- Includes a new set of unpackaged metadata for deploying into a managed LDV org to add tabs and other configuration

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: datasets/rd2/readme.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Using the RD SnowFakery Data Factories
+
+Snowfakery Documentation: https://snowfakery.readthedocs.io/en/docs/#function-blocks
```
Comment (prescod) [inline datasets/rd2/readme.md:3]:
```suggestion
Snowfakery Documentation: https://snowfakery.readthedocs.io/en/docs/
```

**Comment by force2b** [comment]
@prescod Would you mind reviewing this again?

File: cumulusci.yml [hunk 2]
```diff
@@ -655,6 +680,35 @@ tasks:
                 u.LanguageLocaleKey = languageMap.get('%%%PARAM_1%%%');
                 Update u;
 
+    assign_permset:
```
Comment (prescod) [inline cumulusci.yml:683]: I think we can switch to this:

https://cumulusci.readthedocs.io/en/latest/tasks.html#create-permission-set


Especially given this update which will launch next week:

https://github.com/SFDO-Tooling/CumulusCI/pull/2115

File: datasets/rd2/data_recipe_with_rd2.yml [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+      ## 95% of the Contacts will have RD and 5% will not
+      count:
+        random_choice:
+          - choice:
+              probability: 95%
+              pick: 1
+          - choice:
+              probability: 5%
+              pick: 0
+      ## include: RD_RandomDonor, RD_RandomSchedule
```
Comment (prescod) [inline datasets/rd2/data_recipe_with_rd2.yml:103]: Did you want to turn this on now that macros within macros are allowed?

File: unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (prescod) [inline unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: This is fine but just so you know that there is also another way:

https://cumulusci.readthedocs.io/en/latest/tasks.html#create-bulk-data-permission-set

File: datasets/rd2/data_recipe_with_rd2.yml [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    Related_Opportunity_Contact_Role__c:
+      random_choice:
+        - choice:
+            probability: 50%
+            pick: Soft Credit
+        - choice:
+            probability: 50%
+            pick:
+
+## Generate a 10 households that has two household members and a single closed won opportunity
```
Comment (prescod) [inline datasets/rd2/data_recipe_with_rd2.yml:298]:
```suggestion
## Generate 10 households that has two household members and a single closed won opportunity
```

File: cumulusci.yml (hunk 2 above)
Comment (force2b) [reply cumulusci.yml:683]: Very nice, especially `create_bulk_data_permission_set`.

File: datasets/rd2/data_recipe_with_rd2.yml (hunk 3 above)
Comment (force2b) [reply datasets/rd2/data_recipe_with_rd2.yml:103]: Definitely. I'll test it out shortly.

File: unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset (hunk 4 above)
Comment (force2b) [reply unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: Didn't know that existed until now.

**Comment by prescod** [comment]
Things are always changing in CCI world! Can be hard to keep up.

File: unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset (hunk 4 above)
Comment (force2b) [reply unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: @prescod  I just tried this new task and got the following error:
```
Error: Malformed request https://enterprise-saas-32582-dev-ed.cs68.my.salesforce.com/services/data/v48.0/sobjects/PermissionSet/. Response content: [{'message': "No such column 'PermissionsCreateAuditFields' on sobject of type PermissionSet", 'errorCode': 'INVALID_FIELD'}]
```

Comment (prescod) [reply unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: I didn't read the docs before I linked to them. They say: "NOTE: the org setting to allow Set Audit Fields must be turned on." You can decide for yourself whether that makes it more trouble than its worth in this context.

Comment (force2b) [reply unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: Ah. I'm pretty sure that has to be manually enabled in an org. I'll try using `create_permission_set` and specifying the perm name.

Comment (force2b) [reply unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: That worked perfectly. Definitely better than having to include a perm set in unpackaged.

Comment (force2b) [reply unpackaged/config/config_ldv_org_for_testing/permissionsets/HardDelete.permissionset:1]: Did I read that the `create_permission_set` command is going to first check for the existence of the perm set before it tries to recreate?

**Comment by force2b** [comment]
@prescod Do you have any concerns with me merging this into master? I'd to make the data generator available to everyone and all branches. If not, would you mind doing a review? I did delete a remove of the RD2 managed tasks/flows in this since they're no longer needed, but I can revert those if you prefer.

**Comment by prescod** [comment]
@force2b  I have no concerns about merging these!

**Comment by prescod** [comment]
From a process point of view I would prefer that you also get a review from someone on the NPSP team that indicates that all of the metadata changes are harmless. I trust that they are but I shouldn't make assertions outside of my area of expertise.

**Review by prescod — APPROVED** [review]

**Comment by prescod** [comment]
Also: please don't delete the branch until I get a chance to double check whether anything in MetaCI is relying on it. It's possible to merge without deleting branches.

**Comment by force2b** [comment]
@codexmarcel Would you mind doing a review of this to make sure I haven't broken anything specific to NPSP metadata flows. 
@davidmreed I made a change to the yml to remove the 'deploy_rd2_config_managed` task since it shouldn't be needed anymore, and updated the flow used by MetaDeploy. Do you foresee any issues with that change?

File: cumulusci.yml [hunk 6]
```diff
@@ -926,25 +912,6 @@ flows:
                 options:
                     class_name: RD2_DataMigration_BATCH
 
-    enable_rd2_managed:
```
Comment (davidmreed) [inline cumulusci.yml:929]: Do we know if any downstream or Robot automation is still using this flow?

File: cumulusci.yml [hunk 7]
```diff
@@ -1207,6 +1174,113 @@ flows:
...
+                        - CRLP_Rollup_TDTM
+                        - RLLP_OppRollup_TDTM
+                        - RD2_RecurringDonationsOpp_TDTM
+                    active: false
+                    restore: false
+            15:
+                task: unschedule_apex
+                ## Delete all scheduled NPSP jobs ... just in case.
+            16:
+                task: set_duplicate_rule_status
```
Comment (davidmreed) [inline cumulusci.yml:1202]: I love this sequence - it's an awesome example of using OOTB tools to achieve a complex objective and I'm definitely going to steal it for a presentation sometime.

File: datasets/rd2/config_npsp_for_ldv_data_load.cls [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+public static void before_data_load() {
+
+    npe03__Recurring_Donations_Settings__c rdSettings = npe03__Recurring_Donations_Settings__c.getOrgDefaults();
+    npe01__Contacts_And_Orgs_Settings__c cAndOSettings = npe01__Contacts_And_Orgs_Settings__c.getOrgDefaults();
+    %%%NAMESPACE%%%Household_Naming_Settings__c namingSettings = %%%NAMESPACE%%%Household_Naming_Settings__c.getOrgDefaults();
+
+    if (rdSettings.%%%NAMESPACE%%%IsRecurringDonations2Enabled__c == false) {
+        String taskName = 'enable_rd2';
+        if ('%%%NAMESPACE%%%' != '') {
+            taskName += '_managed';
```
Comment (davidmreed) [inline datasets/rd2/config_npsp_for_ldv_data_load.cls:28]: We're removing this flow, right?

File: datasets/rd2/config_npsp_for_ldv_data_load.cls [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+ * @description Before the data load:
+ * - Validate the RD2 is enabled in the org
+ * - Set "First Installment" to never to allow bulk data loading without impacting cpu
+ * - Configure automatic naming so that all new RD's are named nicely
+ * - Mock the migration just to avoid any validation errors
+ * - Configuring the Household Naming format as "{Lastname} {FirstName} Household" so that it's easier to.
+ *   read when there are duplicate households with the same last name.
+ * - Disable Household Address auto creation
+ */
+public static void before_data_load() {
```
Comment (davidmreed) [inline datasets/rd2/config_npsp_for_ldv_data_load.cls:19]: I will add a Work Item to the backlog to have our `load_custom_settings` task support namespace injection, which would obviate the need for Anonymous Apex classes like this.

File: documentation/automation.md [hunk 10]
```diff
@@ -148,6 +146,15 @@ Cumulus's unpackaged configuration is used as follows:
 
 ## Data Sets
 
+### Data Cleanup Tasks
+
+| Name                          | Type | Description                                                                         |
+| ----------------------------- | ---- | ----------------------------------------------------------------------------------- |
+| `test_data_delete` | Task | Delete all records NPSP related data including Opportunity, Account, Contact, RecurringDonation, Level, Engagement Plan, Data Import and Errors |
+| `test_data_delete_managed` | Task | Version for Managed Orgs  |
```
Comment (davidmreed) [inline documentation/automation.md:154]: This task is removed

File: cumulusci.yml (hunk 6 above)
Comment (force2b) [reply cumulusci.yml:929]: Great question, and I think when I make these changes initially I didn't realize that Robot could call a flow this way. The `npsp_robot` file does explicitly call this. If _managed_ flows are no longer needed, I'd prefer to fix the robot file accordingly. However, that [robot file actually goes a lot deeper](https://github.com/SalesforceFoundation/NPSP/blob/master/robot/Cumulus/resources/NPSP.robot#L431). If unmanaged, it looks like it's recreated the entire `enable_rd` flow in robot. I'm not really sure why that is the case though. I should just be able to call the one flow regardless of managed or unmanaged. 
@rjanjanam What do you think?

File: datasets/rd2/config_npsp_for_ldv_data_load.cls (hunk 8 above)
Comment (force2b) [reply datasets/rd2/config_npsp_for_ldv_data_load.cls:28]: Oops. Great catch

File: cumulusci.yml (hunk 6 above)
Comment (force2b) [reply cumulusci.yml:929]: However, that may be more than I want to take on in this PR. I think I'll revert this one change and we can address the use of `enable_rd2` vs. `enabled_rd2_managed` separately.

Comment (rjanjanam) [reply cumulusci.yml]: That is great if one flow can serve for both. I am address in my next PR if you want to make that change in NPSP.robot. Currently I am checking the namespace and the backend call to ensure RD2 is enabled or not and running the flow accordingly. So I am needing both. I had to implement the entire blow for QA back then as I did not know I can call it from Run process.
```    [Documentation]           Checks if Rd2 settings are already enabled and then run the scripts to enable RD2
    ${is_rd2_enabled} =       API Query Recurrring Donation Settings For RD2 Enablement
    ${ns} =                   Get Npsp Namespace Prefix
    Log To Console             ${ns}
    Run Keyword if            "${is_rd2_enabled}"!="True" and "${ns}"!="npsp__"
    ...                       Enable RD2QA
    Run Keyword if            "${is_rd2_enabled}"!="True" and "${ns}"=="npsp__"
    ...                       Run Process     cci  flow  run  enable_rd2_managed

**Review by davidmreed — APPROVED** [review]

File: cumulusci.yml (hunk 6 above)
Comment (force2b) [reply cumulusci.yml]: Thank @rjanjanam. The one flow does work fine, at least when running from the command line. However, before I remove the _managed_ flow from the yml, I'd feel more comfortable knowing that you've updated `npsp.robot` to confirm that just calling the `enable_rd2` directly works fine for both managed and unmanaged in a robot context.

**Pull request merged**